File: internal/schedule/presets.go
package schedule

import (
	"strconv"
	"strings"

	"github.com/mantonx/loomarr/internal/provision"
)

// Preset lowering (programming-design §6.6): the closed authoring vocabulary. The LLM/UI
// compose rules from tokens; these pure functions lower each token to concrete
// SchedulingRule fields. An unknown token returns ok=false so the caller drops it (a rule
// that loses all its parts degrades to the base policy — the §8 fail-safe). No token here
// emits a raw WhenPredicate the model could have hand-authored: the model picks a name, the
// table supplies the predicate.
//
// The suggester's groundPolicy calls these and then applies the grounding CLAMPS (series
// intersection, stricter-only audience ceiling, window bound) — those need the channel's
// grounded picks/ceiling, which live in the suggester, so they are not in this pure table.

// LowerWhen maps a WHEN token to a (predicate, defaultPriority, ok). Composite tokens like
// "weekend-mornings" AND their parts. Case-insensitive. "holiday:<id>" carries the id.
func LowerWhen(token string) (WhenPredicate, int, bool) {
	t := strings.ToLower(strings.TrimSpace(token))

	// holiday:<id> — the §6 calendar. Validate the id against the builtin calendar.
	if id, ok := strings.CutPrefix(t, "holiday:"); ok {
		id = strings.TrimSpace(id)
		if !knownHoliday(id) {
			return WhenPredicate{}, 0, false
		}
		return WhenPredicate{Holiday: id}, 60, true
	}

	// Composite "a-b" (e.g. weekend-mornings) = AND of two base tokens, priority = max+5.
	if base, sub, found := strings.Cut(t, "-"); found {
		if w1, p1, ok1 := lowerBaseWhen(base); ok1 {
			if w2, p2, ok2 := lowerBaseWhen(sub); ok2 {
				return andPredicate(w1, w2), maxInt(p1, p2) + 5, true
			}
		}
		// fall through: "early-morning"/"late-night"/"graveyard" are single base tokens.
	}
	return lowerBaseWhen(t)
}

// lowerBaseWhen handles the atomic (non-composite, non-holiday) WHEN tokens.
func lowerBaseWhen(t string) (WhenPredicate, int, bool) {
	switch t {
	case "weekend":
		return WhenPredicate{Weekend: true}, 20, true
	case "weekday":
		return WhenPredicate{Weekday: true}, 20, true
	case "mornings", "early morning", "early-morning", "morning":
		return WhenPredicate{HourFrom: 6, HourTo: 10}, 30, true
	case "daytime":
		return WhenPredicate{HourFrom: 10, HourTo: 17}, 30, true
	case "primetime", "prime-time", "prime time":
		return WhenPredicate{HourFrom: 20, HourTo: 23}, 40, true
	case "late-night", "late night", "latenight":
		return WhenPredicate{HourFrom: 23, HourTo: 2}, 40, true // wraps
	case "overnight", "graveyard":
		return WhenPredicate{HourFrom: 2, HourTo: 6}, 40, true
	default:
		return WhenPredicate{}, 0, false
	}
}

// andPredicate combines two sub-predicates (used for composite WHEN tokens). Only the
// day-kind + hour-window fields compose here (the base tokens set exactly those).
func andPredicate(a, b WhenPredicate) WhenPredicate {
	out := a
	if b.Weekend {
		out.Weekend = true
	}
	if b.Weekday {
		out.Weekday = true
	}
	if b.HourFrom != 0 || b.HourTo != 0 {
		out.HourFrom, out.HourTo = b.HourFrom, b.HourTo
	}
	return out
}

// LowerWhat maps a WHAT token to a *ScopePolicy that only NARROWS. Returns (nil, true) for
// "all" (the base — no narrowing) and (scope, true) for a recognized narrowing token.
// (scope, true) with a needsCeiling hint for kids/family so the caller applies the
// stricter-only audience clamp. Unknown → (nil, false), dropped by the caller.
//
// A `series:<key>` scope is returned as-is; groundPolicy intersects it with grounded picks.
func LowerWhat(token string) (scope *ScopePolicy, kidsCeiling Rating, ok bool) {
	t := strings.ToLower(strings.TrimSpace(token))
	switch {
	case t == "all", t == "":
		return nil, "", true // no narrowing (the base)
	case t == "holiday-matched":
		// Seasonal keyword filter is applied by the seasonal engine, not scope; a rule
		// signals it by carrying a nil scope + relying on the channel seasonal mode. v1:
		// treat as "no extra scope" here; the holiday WHEN already gates the timing.
		return nil, "", true
	case t == "kids":
		return &ScopePolicy{Genres: GenreFilter{Include: kidSafeGenres}}, NormalizeRating("TV-Y7"), true
	case t == "family":
		return &ScopePolicy{Genres: GenreFilter{Include: familyGenres}}, NormalizeRating("TV-PG"), true
	case strings.HasPrefix(t, "series:"):
		key := strings.TrimSpace(token[len("series:"):]) // preserve original case for the key
		if key == "" {
			return nil, "", false
		}
		return &ScopePolicy{Series: []provision.Key{provision.Key(key)}}, "", true
	case strings.HasPrefix(t, "genre-not:"):
		g := strings.TrimSpace(token[len("genre-not:"):])
		if g == "" {
			return nil, "", false
		}
		return &ScopePolicy{Genres: GenreFilter{Exclude: []string{g}}}, "", true
	case strings.HasPrefix(t, "genre:"):
		g := strings.TrimSpace(token[len("genre:"):])
		if g == "" {
			return nil, "", false
		}
		return &ScopePolicy{Genres: GenreFilter{Include: []string{g}}}, "", true
	case strings.HasPrefix(t, "era:"):
		if r, ok := parseEraToken(t[len("era:"):]); ok {
			return &ScopePolicy{Era: r}, "", true
		}
		return nil, "", false
	default:
		return nil, "", false
	}
}

// LowerHow maps a HOW token to a RuleOrdering + a per-rule Window (0 = inherit). Unknown →
// (zero, 0, false), dropped by the caller (the rule then inherits the channel ordering).
func LowerHow(token string) (RuleOrdering, Duration, bool) {
	switch strings.ToLower(strings.TrimSpace(token)) {
	case "syndication":
		return RuleOrdering{Ordering: OrderSyndication}, 0, true
	case "shuffle":
		return RuleOrdering{Ordering: OrderShuffle}, 0, true
	case "marathon":
		// sequential + unbounded block (binge one show) + no breaks + don't truncate a binge.
		// BlockMax -1, NOT 0: applyRuleHow overlays a Separation field only when it's != 0
		// (0 means "inherit the channel's blockMax"), while the separation engine reads any
		// value <= 0 as UNBOUNDED. -1 is the value that is both non-zero (so it overlays) and
		// <= 0 (so it disables the block cap) — letting one series binge without interruption.
		return RuleOrdering{
			Ordering:   OrderSequential,
			Separation: SeparationPolicy{BlockMax: -1},
			NoBreaks:   true,
		}, WindowFull, true
	case "feature", "sequential":
		return RuleOrdering{Ordering: OrderSequential}, 0, true
	default:
		return RuleOrdering{}, 0, false
	}
}

// kidSafeGenres / familyGenres are the include-sets for the kids/family WHAT tokens. These
// are display genre names matched case-insensitively at enforcement (§4); an unknown name
// simply never matches (harmless). The audience CEILING (returned alongside) is the real
// safety gate — genres are a soft preference, the ceiling is the hard floor.
var (
	kidSafeGenres = []string{"Animation", "Family", "Kids"}
	familyGenres  = []string{"Family", "Animation", "Adventure", "Comedy"}
)

// knownHoliday reports whether id is a builtin-calendar holiday (§6), so an unknown
// "holiday:xyz" token is dropped rather than lowered to a never-matching predicate.
func knownHoliday(id string) bool {
	for _, h := range builtinCalendar {
		if h.id == id {
			return true
		}
	}
	return false
}

// parseEraToken parses "1990-1999" / "1990-" / "-1999" into a *Range. Both bounds optional
// but at least one required. Returns (nil, false) on garbage.
func parseEraToken(s string) (*Range, bool) {
	from, to, found := strings.Cut(strings.TrimSpace(s), "-")
	if !found {
		return nil, false
	}
	r := &Range{}
	if from = strings.TrimSpace(from); from != "" {
		n, err := strconv.Atoi(from)
		if err != nil || n <= 0 {
			return nil, false
		}
		r.From = n
	}
	if to = strings.TrimSpace(to); to != "" {
		n, err := strconv.Atoi(to)
		if err != nil || n <= 0 {
			return nil, false
		}
		r.To = n
	}
	if r.From == 0 && r.To == 0 {
		return nil, false
	}
	return r, true
}

func maxInt(a, b int) int {
	if a > b {
		return a
	}
	return b
}

File: web/apps/web/src/components/loomarr/channel-rules-editor/presets/presets.ts
import type { RuleOrdering, ScopePolicy, WhenPredicate } from "@loomarr/api";

// The FE mirror of internal/schedule/presets.go (programming-design §6.6 — the closed
// authoring vocabulary). Keep this table BYTE-FOR-BYTE in sync with presets.go: same
// tokens, same predicates/scopes/orderings, same default priorities. It exists so a
// hand-authored rule in this editor lowers to the identical SchedulingRule shape an
// LLM-authored one would (§8.1) — a UX convenience for building the picker + a live
// preview label, NOT the source of truth. The backend re-validates on every write
// (unknown tokens dropped, window clamped, audience stricter-only, §6.6 grounding) —
// so a bug here degrades to "the picker offered something the server then dropped",
// never a safety hole. If presets.go changes, update this file in the same PR.

// ---- WHEN tokens (WhenPredicate) ----------------------------------------------------

interface WhenPreset {
  token: string;
  label: string;
  // A terser form of `label` for the computed row caption (label.ts) — the Select's own
  // item text carries the hour-range explanation, so the caption doesn't repeat it.
  shortLabel: string;
  predicate: WhenPredicate;
  priority: number;
}

// Holiday ids match the §6 builtinCalendar (knownHoliday in presets.go). Kept as a
// small closed list here too — the BE is the real gate, this is just what the picker
// offers.
const HOLIDAY_IDS = ["christmas", "halloween", "thanksgiving", "newyear", "valentines"] as const;

const HOLIDAY_LABELS: Record<(typeof HOLIDAY_IDS)[number], string> = {
  christmas: "Christmas",
  halloween: "Halloween",
  thanksgiving: "Thanksgiving",
  newyear: "New Year",
  valentines: "Valentine's Day",
};

// The base (non-holiday, non-composite) WHEN tokens — mirrors lowerBaseWhen exactly,
// including its default priorities (weekend/weekday=20, dayparts=30/40).
const WHEN_PRESETS: WhenPreset[] = [
  { token: "weekend", label: "Weekend", shortLabel: "Weekend", predicate: { weekend: true }, priority: 20 },
  { token: "weekday", label: "Weekday", shortLabel: "Weekday", predicate: { weekday: true }, priority: 20 },
  {
    token: "mornings",
    label: "Mornings (6–10)",
    shortLabel: "Mornings",
    predicate: { hourFrom: 6, hourTo: 10 },
    priority: 30,
  },
  {
    token: "daytime",
    label: "Daytime (10–17)",
    shortLabel: "Daytime",
    predicate: { hourFrom: 10, hourTo: 17 },
    priority: 30,
  },
  {
    token: "primetime",
    label: "Primetime (20–23)",
    shortLabel: "Primetime",
    predicate: { hourFrom: 20, hourTo: 23 },
    priority: 40,
  },
  {
    token: "late-night",
    label: "Late night (23–2)",
    shortLabel: "Late night",
    predicate: { hourFrom: 23, hourTo: 2 },
    priority: 40,
  },
  {
    token: "overnight",
    label: "Overnight (2–6)",
    shortLabel: "Overnight",
    predicate: { hourFrom: 2, hourTo: 6 },
    priority: 40,
  },
  ...HOLIDAY_IDS.map((id) => ({
    token: `holiday:${id}`,
    label: HOLIDAY_LABELS[id],
    shortLabel: HOLIDAY_LABELS[id],
    predicate: { holiday: id },
    priority: 60,
  })),
];

// The options the WHEN select offers, grouped only visually — the underlying value is
// always the flat token string.
const WHEN_OPTIONS: { value: string; label: string }[] = WHEN_PRESETS.map((p) => ({
  value: p.token,
  label: p.label,
}));

// WHEN_SHORT_LABELS — token → terse label, for the computed row caption (label.ts).
const WHEN_SHORT_LABELS: { value: string; label: string }[] = WHEN_PRESETS.map((p) => ({
  value: p.token,
  label: p.shortLabel,
}));

// lowerWhen — the FE mirror of LowerWhen (presets.go). No composite ("weekend-mornings")
// support here: the editor offers only the atomic tokens (composites are an LLM/authoring
// convenience the BE still accepts, but the picker doesn't need to compose them itself).
const lowerWhen = (token: string): { predicate: WhenPredicate; priority: number } | undefined => {
  const preset = WHEN_PRESETS.find((p) => p.token === token);
  if (!preset) return undefined;
  return { predicate: preset.predicate, priority: preset.priority };
};

// tokenForWhen — the reverse lookup: given an existing rule's WhenPredicate, find which
// preset token (if any) produced it, so the editor can select the right option when
// loading an existing rule. A predicate that isn't a byte-for-byte match to a known
// preset (e.g. hand-authored via the API, or a composite) falls back to "" (shown as
// "Custom" in the select — still saved as-is since the editor only ever REPLACES the
// token's own fields, never the whole rule from scratch, when a preset can't be matched).
const tokenForWhen = (w: WhenPredicate | undefined): string => {
  if (!w) return "";
  const match = WHEN_PRESETS.find(
    (p) =>
      Boolean(p.predicate.weekend) === Boolean(w.weekend) &&
      Boolean(p.predicate.weekday) === Boolean(w.weekday) &&
      (p.predicate.hourFrom ?? 0) === (w.hourFrom ?? 0) &&
      (p.predicate.hourTo ?? 0) === (w.hourTo ?? 0) &&
      (p.predicate.holiday ?? "") === (w.holiday ?? ""),
  );
  return match?.token ?? "";
};

// ---- WHAT tokens (ScopePolicy) -------------------------------------------------------

type WhatKind = "all" | "series" | "genre" | "kids" | "family" | "holiday-matched" | "era";

interface WhatOption {
  value: string;
  label: string;
}

// Static WHAT options (the ones that don't need the channel's lineup). `series:<key>` is
// appended by the caller from `lineupKeys` (§6.6: "intersected with the channel's
// actually-grounded picks").
const WHAT_STATIC_OPTIONS: WhatOption[] = [
  { value: "all", label: "Anything (no extra narrowing)" },
  { value: "kids", label: "Kids-safe genres" },
  { value: "family", label: "Family genres" },
  { value: "holiday-matched", label: "Holiday-matched titles" },
];

// lowerWhat — the FE mirror of LowerWhat (presets.go). Returns undefined scope for "all"
// (no narrowing) and a ScopePolicy for a recognized narrowing token. Unknown → null (drop).
const lowerWhat = (token: string): { scope: ScopePolicy | undefined } | undefined => {
  const t = token.trim();
  if (t === "" || t === "all") return { scope: undefined };
  if (t === "kids") return { scope: { genres: { include: ["Animation", "Family", "Kids"] } } };
  if (t === "family")
    return { scope: { genres: { include: ["Family", "Animation", "Adventure", "Comedy"] } } };
  if (t === "holiday-matched") return { scope: undefined }; // seasonal engine handles the filter, not scope
  if (t.startsWith("series:")) {
    const key = t.slice("series:".length).trim();
    if (!key) return undefined;
    return { scope: { series: [key] } };
  }
  if (t.startsWith("genre:")) {
    const g = t.slice("genre:".length).trim();
    if (!g) return undefined;
    return { scope: { genres: { include: [g] } } };
  }
  if (t.startsWith("era:")) {
    const parsed = parseEraToken(t.slice("era:".length));
    if (!parsed) return undefined;
    return { scope: { era: parsed } };
  }
  return undefined;
};

// tokenForWhat — the reverse lookup for an existing rule's ScopePolicy, so the editor can
// preselect the WHAT option when loading a rule. Falls back to "all" for a nil/empty
// scope and "" ("Custom") for a scope shape none of the presets produce.
const tokenForWhat = (scope: ScopePolicy | null | undefined): string => {
  if (!scope) return "all";
  const series = scope.series;
  if (series && series.length > 0) return `series:${series[0]}`;
  const include = scope.genres?.include ?? [];
  if (arraysEqual(include, ["Animation", "Family", "Kids"])) return "kids";
  if (arraysEqual(include, ["Family", "Animation", "Adventure", "Comedy"])) return "family";
  if (include.length === 1 && !scope.genres?.exclude?.length && !scope.era && !scope.seasons) {
    return `genre:${include[0]}`;
  }
  if (scope.era && (scope.era.from || scope.era.to)) {
    return `era:${scope.era.from ?? ""}-${scope.era.to ?? ""}`;
  }
  return "";
};

const arraysEqual = (a: string[], b: string[]): boolean =>
  a.length === b.length && a.every((v, i) => v === b[i]);

// parseEraToken — mirrors parseEraToken (presets.go): "1990-1999" / "1990-" / "-1999".
const parseEraToken = (s: string): { from?: number; to?: number } | undefined => {
  const trimmed = s.trim();
  const idx = trimmed.indexOf("-");
  if (idx === -1) return undefined;
  const fromStr = trimmed.slice(0, idx).trim();
  const toStr = trimmed.slice(idx + 1).trim();
  const range: { from?: number; to?: number } = {};
  if (fromStr !== "") {
    const n = Number(fromStr);
    if (!Number.isInteger(n) || n <= 0) return undefined;
    range.from = n;
  }
  if (toStr !== "") {
    const n = Number(toStr);
    if (!Number.isInteger(n) || n <= 0) return undefined;
    range.to = n;
  }
  if (range.from === undefined && range.to === undefined) return undefined;
  return range;
};

// ---- HOW tokens (RuleOrdering + Window) ----------------------------------------------

interface HowPreset {
  token: string;
  label: string;
  // A terser form of `label` for the computed row caption (label.ts) — the Select's own
  // item text carries the parenthetical explanation, so the caption doesn't repeat it.
  shortLabel: string;
  ordering: RuleOrdering;
  // "full" = the WindowFull sentinel (marathon: don't truncate a binge); undefined =
  // inherit the channel/global window.
  window?: "full";
}

// WindowFull sentinel — mirrors schedule.WindowFull (-1) on the Go side. The wire type
// for ChannelPolicy/SchedulingRule.window is a Go-duration STRING; there is no numeric
// "-1" on this side of the wire, so the editor tracks "full run" as a distinct sentinel
// string and the caller (channel-rules-editor.tsx) is responsible for putting it on
// `rule.window` in whatever string form the rest of the app already uses for "unbounded"
// (kept out of this pure lowering table, which only knows tokens ⇄ RuleOrdering).
const HOW_PRESETS: HowPreset[] = [
  {
    token: "syndication",
    label: "Syndication (the deck order)",
    shortLabel: "Syndication",
    ordering: { ordering: "syndication" },
  },
  { token: "shuffle", label: "Shuffle", shortLabel: "Shuffle", ordering: { ordering: "shuffle" } },
  {
    token: "marathon",
    label: "Marathon (binge one show, no breaks)",
    shortLabel: "Marathon",
    ordering: { ordering: "sequential", noBreaks: true, separation: { blockMax: -1 } },
    window: "full",
  },
  {
    token: "feature",
    label: "Feature (in order)",
    shortLabel: "Feature",
    ordering: { ordering: "sequential" },
  },
];

const HOW_OPTIONS: { value: string; label: string }[] = HOW_PRESETS.map((p) => ({
  value: p.token,
  label: p.label,
}));

// HOW_SHORT_LABELS — token → terse label, for the computed row caption (label.ts).
const HOW_SHORT_LABELS: { value: string; label: string }[] = HOW_PRESETS.map((p) => ({
  value: p.token,
  label: p.shortLabel,
}));

// lowerHow — the FE mirror of LowerHow (presets.go). "sequential" is accepted as an
// alias for "feature" on the Go side but the editor only ever emits the canonical token.
const lowerHow = (token: string): { ordering: RuleOrdering; window?: "full" } | undefined => {
  const preset = HOW_PRESETS.find((p) => p.token === token);
  if (!preset) return undefined;
  return { ordering: preset.ordering, window: preset.window };
};

// tokenForHow — the reverse lookup for an existing rule's RuleOrdering, so the editor can
// preselect the HOW option when loading a rule. "" ("Custom") when nothing matches.
const tokenForHow = (how: RuleOrdering | undefined): string => {
  if (!how?.ordering) return "";
  const match = HOW_PRESETS.find(
    (p) =>
      p.ordering.ordering === how.ordering &&
      Boolean(p.ordering.noBreaks) === Boolean(how.noBreaks) &&
      (p.ordering.separation?.blockMax ?? 0) === (how.separation?.blockMax ?? 0),
  );
  return match?.token ?? "";
};

export type { WhatKind };
export {
  HOLIDAY_IDS,
  HOW_OPTIONS,
  HOW_SHORT_LABELS,
  lowerHow,
  lowerWhat,
  lowerWhen,
  tokenForHow,
  tokenForWhat,
  tokenForWhen,
  WHAT_STATIC_OPTIONS,
  WHEN_OPTIONS,
  WHEN_SHORT_LABELS,
};
